libraries/Microsoft.Bot.Builder.Core.State/StateManager.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder.Core.State
+{
+    public class StateManager : IStateManager
+    {
+        private readonly Dictionary<string, StateEntry> _state = new Dictionary<string, StateEntry>();
+
+        public StateManager(string stateNamespace, IStateStorageProvider storageProvider)
+        {
+            Namespace = stateNamespace ?? throw new ArgumentNullException(nameof(stateNamespace));
+            StorageProvider = storageProvider ?? throw new ArgumentNullException(nameof(storageProvider));
+        }
+
+        public string Namespace { get; }
+
+        public IStateStorageProvider StorageProvider { get; }
+
+        public async Task<TState> Get<TState>(string key) where TState : class, new()
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (!_state.TryGetValue(key, out var stateEntry))
+            {
+                var stateStorageEntry = await StorageProvider.Load(Namespace, key);
+
+                if (stateStorageEntry == default(IStateStorageEntry))
+                {
+                    return default(TState);
+                }
+
+                _state.Add(key, new StateEntry(stateStorageEntry));
+
+                return stateStorageEntry.GetValue<TState>();
+            }
+
+            return stateEntry.IsPendingDeletion ? default(TState) : stateEntry.StateStorageEntry.GetValue<TState>();
+        }
+
+        public void Set<TState>(string key, TState state) where TState : class
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (!_state.TryGetValue(key, out var stateEntry))
+            {
+                var stateStorageEntry = StorageProvider.CreateNewEntry(Namespace, key);
+
+                stateEntry = new StateEntry(stateStorageEntry, isDirty: true);
+
+                _state.Add(key, stateEntry);
+            }
+            else
+            {
+                stateEntry.IsDirty = true;
+                stateEntry.IsPendingDeletion = false;
+            }
+
+            stateEntry.StateStorageEntry.SetValue(state);
+        }
```

Comment:
I may not be reading this correctly, but I am concerned that 2 components that use the same state object will clobber each other.
A: var x = get(key)
A -> calls B
B: var x get(key)
B: Change x
B: Set(x)
A -> Change x (not the same X)
A: Set(x) <-- clobbers B's change
???

Reply:
_Absolutely_ this is a possibility. Each state namespace, hence state manager, is a shared key space and all that contribute to it must be careful within that space not to step on one another within that space. I honestly don't think this is any different than how some other frameworks out there work honestly. For example, ASP.NET's `HttpContext::Items` collection. 

This is an important thing to call out in documentation and, for middleware especially, provide some guidance for some kind of prefixing scheme to keep their keys isolated.

Couple things though:
 1. One thing to note is that for code that is just using the `Set<T>(T value)` extension method approach, it's actually using the `Type::FullName` of the `T` and, in that case, that's the full namespace qualified name of the .NET type which would all but entirely eliminate any collision concerns.
 2. The absolutely greatest isolation is attainable by using an entirely separate state namespace. That can be done via an explicit namespace constant and via something that I just realized wasn't called out explicitly in this DCR. Developers can actually create their own strongly types state manager interfaces/classes and register those with a `UseState<TStateManager>()` instead. Those kinds of state managers are actually capable of constructing themselves from details of the current `ITurnContext`. This is actually how conversation state and user state (and eventually bot state) work under the covers. This extension point enables you create your own, completely isolated state namespace based off of dynamic values coming `ITurnContext`.